neqo-transport/src/server.rs

```diff
@@ -620,71 +561,68 @@ impl Server {
-    fn process_next_output(&mut self, now: Instant) -> Option<Datagram> {
-        qtrace!([self], "No packet to send, look at waiting connections");
-        while let Some(c) = self.waiting.pop_front() {
-            if let Some(d) = self.process_connection(&c, None, now) {
-                return Some(d);
+    fn process_next_output(&mut self, now: Instant) -> Output {
+        let mut callback = None;
+
+        for connection in self.connections.borrow().values() {
+            match connection.borrow_mut().process(None, now) {
+                Output::None => {}
+                d @ Output::Datagram(_) => return d,
+                Output::Callback(next) => match callback {
+                    Some(previous) => callback = Some(min(previous, next)),
+                    None => callback = Some(next),
+                },
             }
         }
 
-        qtrace!([self], "No packet to send still, check wake up times");
-        loop {
-            let connection = self
-                .connections
-                .borrow()
-                .values()
-                .find(|c| c.borrow().needs_waking(now))
-                .cloned()?;
-            let datagram = self.process_connection(&connection, None, now);
-            connection.borrow_mut().woken();
-            if datagram.is_some() {
-                return datagram;
-            }
-        }
+        callback.map_or(Output::None, Output::Callback)
     }
 
     pub fn process(&mut self, dgram: Option<&Datagram>, now: Instant) -> Output {
-        if self.wake_at.map_or(false, |c| c <= now) {
-            self.wake_at = None;
-        }
+        let out = dgram
+            .map_or(Output::None, |d| self.process_input(d, now))
+            .or_else(|| self.process_next_output(now));
+
+        // Clean-up closed connections.
+        self.connections.borrow_mut().retain(|_, c| {
+            if matches!(c.borrow().state(), State::Closed(_)) {
+                // TODO: Is this still needed?
+                c.borrow_mut().set_qlog(NeqoQlog::disabled());
+                false
+            } else {
+                true
+            }
+        });
 
-        dgram
-            .and_then(|d| self.process_input(d, now))
-            .or_else(|| self.process_next_output(now))
-            .map(|d| {
-                qtrace!([self], "Send packet: {:?}", d);
-                Output::Datagram(d)
-            })
-            .or_else(|| self.wake_at.take().map(|c| Output::Callback(c - now)))
-            .unwrap_or_else(|| {
-                qtrace!([self], "Go dormant");
-                Output::None
-            })
+        out
     }
 
     /// This lists the connections that have received new events
     /// as a result of calling `process()`.
-    pub fn active_connections(&mut self) -> Vec<ActiveConnectionRef> {
-        mem::take(&mut self.active).into_iter().collect()
+    pub fn active_connections(&mut self) -> HashSet<ActiveConnectionRef> {
```

Comment:
Should this return `impl Iterator<Item = ActiveConnectionRef>` instead?  That way we save on the allocation.  If code is doing - as you do in tests - very little with the iterator, then this won't be a problem.  But they can call `.collect()` if they are looking to compare notes or anything like that.

Reply:
`self.connections` can contain multiple references to the same `Connection`. `ActiveConnectionRef` `Hash` implementation makes sure two references to the same `Connection` have the same hash:

https://github.com/mozilla/neqo/blob/ea542735006a1075f97a6984d86029d9b5126d6a/neqo-transport/src/server.rs#L712-L723

The `collect()` into a `HashSet` serves the purpose of de-duplication here.

Note that this pull request does not introduce a new allocation per `active_connections` call. The previous implementation also allocated into a `HashSet`, see `self.active` being set back to `HashSet::default` in the above `mem::take`.

We could still return an `impl Iterator` from the `HashSet`. That said, `neqo-http3` calling `active_connections` again needs a `HashSet`.

Does that make sense @martinthomson?

Reply:
Ah, I see.  That's fine.  Maybe a comment to that effect would help (I wouldn't want to duplicate that comment though, so ignore at your discretion).